### DISCUSSION TITLE
Are there any standard unit test batteries for the standard interfaces? (IDictionary<K,V>, etc)

### DISCUSSION THREAD

**Discussion by daiplusplus** [dotnet/runtime#52528] [Q&A]
(I would have posted this to StackOverflow but as I'm asking for a resource or recommendation this question would violate SO's rules against posts asking for recommendations and resources)

I found myself implementing `IDictionary<TKey,TValue>` (`IDictionary` for short, I'm not referring to the legacy non-generic interface) and as part of TDD I need to write unit tests for my implementation, and it wouldn't be a good use of my time to write my own unit tests for my implementation of `IDictionary` as it would be identical to any existing test for the standard functionality of `IDictionary`, so I searched around for some standard, reusable test-cases for `IDictionary`, but I couldn't find anything like a copy+paste battery of pre-written unit-tests or NuGet package with the test-cases built-in. The closest I found [was this, but seems specific to someone's own projects](https://www.nuget.org/packages/ReadonlyDictionaryTests/).

I took a look at the official .NET tests for the generic collections [but they seem to be for their specific concrete types, rather than reusable tests for each collection interface](https://github.com/dotnet/runtime/tree/main/src/libraries/System.Collections/tests/Generic) - and this obviously doesn't concern the non-collection interfaces that should also be tested when reimplemented, such as `IEquatable<T>` and `IComparer<T>`.

What do you do when you need to unit-test types implementing standard interfaces?

**Comment by nd1012** [comment]
Right now I'm working at it - but indeed I'm not a test code writer :( I'm still looking for Open Source Unit tests for those standard interfaces.

Anyway, how/why would you implement tests for those? Most of the interfaces are generic, and depending on the tests you would require a specialized test data set which can fit the tests requirements. For that reason I wrote only very basic tests which don't rely on the test data too much.

I think that's maybe the way to go: Use only basic tests for interfaces and write specialized tests for types. However, I couldn't even find basic tests to adopt for interfaces like `IDictionary` - that's why I decided to write my own today.

Maybe you can get some ideas from [my repository](https://github.com/nd1012/wan24-Tests). The tests I wrote are in the `dev` branch for now - I'm looking forward to merge them to `main` and publish a new [NuGet package](https://www.nuget.org/packages/wan24-Tests) within this weekend.
